### ISSUE TITLE
`uv python pin` should find `.python-version` file in a workspace

### ISSUE THREAD

**Issue by zanieb** [astral-sh/uv#4971]
If working in a child directory of a workspace, `uv python pin` should support finding the `.python-version` file in the workspace root.

Related #4970

**Comment by silvanocerza** [comment]
I thought of changing [`version_files.rs`](https://github.com/astral-sh/uv/blob/main/crates/uv-python/src/version_files.rs) to use [`Workspace::discovery()`](https://github.com/astral-sh/uv/blob/8d6c49b36c5ac70729274e62fd9f09eb760ed4c0/crates/uv-distribution/src/workspace.rs#L70) internally to find the version files. Though `uv-distribution` can't be added as a dependency of `uv-python` as that would creation a cycle.

Should I instead change all functions in `version_files.rs` to accept a path as argument and search the version files there? Or is there a simpler solution that I'm missing cause of lack of context?

**Comment by zanieb** [comment]
Yeah based on that description I'd probably add `requests_from_version_file_in(root: PathBuf)` and keep the workspace discovery over in the `uv` crate.

Thanks for looking into it!

**Comment by j178** [comment]
Should `uv python pin 3.12` also only write to the project root or the workspace root? At present, it writes to the current directory, but this means it won't be picked up by other commands like `uv python install` unless they're run in that specific directory.

**Comment by zanieb** [comment]
This is a good (and open) question. Do users opt-in to operating on the project or operating on the workspace? When working in the workspace root, we require `--project <name>` to affect a project. When in a project directory, we should probably require `--workspace` to mutate the workspace?

**Comment by silvanocerza** [comment]
I'm assuming that a workspace can contain more projects, correct me if I'm wrong. Given that I would image this behaviour.

If called in a project subfolder it would pin for the first closest project.
If called in a workspace folder it would pin for the workspace only.
If you want to change this behaviour you'd use the `--project <name>` or `--workspace` options as you propose.
The `--isolate` option ignores all the above and pins in the current working directory as you said in https://github.com/astral-sh/uv/pull/5035#issuecomment-2226960463.

It makes sense in my head but maybe it could be confusing to some? 🤔
